### PR TITLE
Added events documentation

### PR DESCRIPTION
**Pull request by tanmaymunjal**

### CODE REVIEW

File: src/anchor_in_depth/Events.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+Events in Anchor Solana provide a powerful mechanism for notifying and communicating between different components of a Solana dApp. They allow for the emission and tracking of significant occurrences within the program's execution. This documentation will cover the concept of events in Anchor Solana, their structure, and how to use them effectively in your smart contract development.
+
+## Table of Contents
+- [Introduction to Events](#introduction-to-events)
+- [Defining Events](#defining-events)
+- [Emitting Events](#emitting-events)
+- [Subscribing to Events](#subscribing-to-events)
+
+## Introduction to Events
+An event is a structured piece of data that holds information about a specific occurrence in a smart contract program. It acts as a message to notify external parties about significant changes or transactions within the contract's state. Events can be used to provide transparency, traceability, and synchronization in decentralized applications.
```
Comment (acheroncrypto) [inline src/anchor_in_depth/Events.md]: > It acts as a message to notify external parties about significant changes or transactions within the contract's state.

The change doesn't need to be significant and contract doesn't have state so it would be better to rephrase this line or delete it.

File: src/anchor_in_depth/Events.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+## Table of Contents
+- [Introduction to Events](#introduction-to-events)
+- [Defining Events](#defining-events)
+- [Emitting Events](#emitting-events)
+- [Subscribing to Events](#subscribing-to-events)
+
+## Introduction to Events
+An event is a structured piece of data that holds information about a specific occurrence in a smart contract program. It acts as a message to notify external parties about significant changes or transactions within the contract's state. Events can be used to provide transparency, traceability, and synchronization in decentralized applications. 
+
+Events in Anchor are basically base64 encoded structs that are program logged. For ex, when you may write msg!() something to debug your Solana program, you are logging data. Events are essentially the same thing but they are not human readable as they are base64 encoded. The reason for this is that it is quite expensive to log string data:
```
Comment (acheroncrypto) [inline src/anchor_in_depth/Events.md]: >  For ex, when you may write msg!() something to debug your Solana program, you are logging data.

Could you rephrase this sentence?

File: src/anchor_in_depth/Events.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    })
+}
+```
+
+## Defining Events
+In Anchor Solana, events are defined using the `#[event]` attribute macro. This macro allows you to specify the fields that an event should contain. Events can include various data types, making them versatile for different use cases.
+
+```rust
+use anchor_lang::prelude::*;
+
```
Comment (acheroncrypto) [inline src/anchor_in_depth/Events.md]:
```suggestion
```

Not needed as snippets are not complete anyway.

File: src/anchor_in_depth/Events.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    from: Pubkey,
+    to: Pubkey,
+    amount: u64,
+}
+```
+
+In this example, we define an event named `TransferEvent` with three fields: `from` (sender's address), `to` (receiver's address), and `amount` (the transferred amount).
+
+## Emitting Events
+To emit an event within your Anchor Solana program, you can use the `emit!` macro. This allows you to create an instance of the event structure and broadcast it to the network.
```
Comment (acheroncrypto) [inline src/anchor_in_depth/Events.md]: > This allows you to create an instance of the event structure and broadcast it to the network.

Using `emit!` does not allow creating an instance of the event structure as they are separate actions.

File: src/anchor_in_depth/Events.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        
+        Ok(())
+    }
+}
+```
+
+In this example, when the `transfer` function is called, a `TransferEvent` is emitted using the `emit!` macro. The relevant data is populated into the event fields.
+
+## Subscribing to Events
+External parties can subscribe to events emitted by your smart contract using Solana's event subscription mechanisms. This enables them to listen for specific events and take actions accordingly.
```
Comment (acheroncrypto) [inline src/anchor_in_depth/Events.md]: It's Anchor's subscription rather than Solana's.

We can also use something like "Anyone" instead of "External parties".

File: src/anchor_in_depth/Events.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+For example, using Anchor ts client library(@project-serum/anchor), you can subscribe to events like this:
```
Comment (acheroncrypto) [inline src/anchor_in_depth/Events.md]: Please use [@coral-xyz/anchor](https://www.npmjs.com/package/@coral-xyz/anchor).

```suggestion
For example, using Anchor TS library([@coral-xyz/anchor](https://www.npmjs.com/package/@coral-xyz/anchor)), you can subscribe to events like this:
```

File: src/anchor_in_depth/Events.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+const program = new anchor.Program(IDL, PROGRAM_ID) as Program<MyProgram>
+
+let listener = null;
+let [event, slot] = await new Promise((resolve, _reject) => {
+  listener = program.addEventListener("TransferEvent", (event, slot) => {
+    resolve([event, slot]);
+  });
+  program.rpc.initialize();
+});
+```
```
Comment (acheroncrypto) [inline src/anchor_in_depth/Events.md]: There is a lot of stuff going on here that takes the focus out of the main important parts which is subscribing to events. It would also be better if we avoid the `Promise` stuff as that's just being used in tests and not realistically how events would be used.

```suggestion
```ts
const subscriptionId = program.addEventListener("TransferEvent", (event) => {
  // Handle event...
});
```
```

**Review by acheroncrypto — COMMENTED** [review]
Hey, thank you for the effort and sorry for the delay.

There are a few things we could mention:

1. Unsubscribe method

```ts
// Unsubscribe to not leak memory
program.removeEventListener(subscriptionId);
```

2. Information about how logs can be truncated by RPC providers therefore making events less reliable

3. CPI Events

**Comment by tanmaymunjal** [comment]
Hey @acheroncrypto, really sorry for the super long delay. Had some life stuff to do and took a break from most coding stuff except what I needed to pay the bills(big drop in GitHub contribs), just started back randomly coding and contributing again!

I made the changes you recommended, feel free to recommend anything else I can to better the doc

**Review by acheroncrypto — APPROVED** [review]
> really sorry for the super long delay

No worries, thank you for making this contribution!

> I made the changes you recommended, feel free to recommend anything else I can to better the doc

Thanks again, recommending the changes would take a lot of time so I've just made the changes 😅
